File: typescript/src/nuc/OBJECT$CLASS.ts
// src/nuc/OBJECT$CLASS.ts

import Instruction from "../Instruction";
import NODE from "./NODE";
import OBJECT$INSTANCE from "./OBJECT$INSTANCE";
import Scope from "../Scope";
import { retrieve } from "../graph";

class OBJECT$CLASS extends NODE {
  class!: {
    instances: Record<string, Record<string, unknown>>;
    declarations: Record<string, OBJECT$CLASS>;
  };
  name!: string;
  type = "CLASS";

  run(scope: Scope): { next: NODE[] } {
    const current = scope.$instance as Record<string, unknown> | undefined;
    const allInstances = current
      ? [current]
      : Object.values(this.class.instances);

    const statements: NODE[] = [];

    for (const inst of allInstances) {
      const instName = (inst as any).name;
      const id = `${instName}.${this.name}`;

      const statement = new OBJECT$INSTANCE(id) as unknown as {
        class?: OBJECT$CLASS["class"];
        object?: Record<string, unknown>;
        name?: string;
      };

      statement.class = this.class;
      statement.object = retrieve(instName);
      statement.name = this.name;

      const instanceScope = new Scope(scope, statement);
      instanceScope.$instance = inst;

      const instr = new Instruction(
        instanceScope,
        statement,
        () => {},
        () => {},
        null,
        null
      ) as unknown as NODE;

      statements.push(instr);
    }

    return { next: statements };
  }

  graph(): void {
    this.class.declarations[this.key] = this;
  }
}

export default OBJECT$CLASS;

File: cpp/src/nuc/OBJECT$CLASS.cpp
#include "nucleoid/nuc/OBJECT$CLASS.hpp"
#include "nucleoid/nuc/OBJECT$INSTANCE.hpp"
#include "nucleoid/Graph.hpp"
#include "nucleoid/Scope.hpp"
#include <cstdint>
#include <random>
#include <sstream>

namespace nucleoid {

static std::string objUUID() {
    static std::mt19937_64 rng(std::random_device{}());
    std::uniform_int_distribution<uint64_t> dist;
    std::ostringstream ss;
    ss << std::hex << dist(rng);
    return ss.str();
}

NODE* OBJECT$CLASS::run(Scope* scope) {
    if (!classRef) return nullptr;

    NODE* inst = scope->instance;
    std::vector<NODE*> instances;
    if (inst) instances.push_back(inst);
    else for (auto& [k, v] : classRef->instances) if (v) instances.push_back(v);

    for (NODE* i : instances) {
        std::string instName = i->key;
        std::string id       = instName + "." + name;

        auto* oi = new OBJECT$INSTANCE(id);
        oi->classNode  = nullptr;
        oi->objectNode = Graph::instance().retrieve(instName);
        oi->name       = name;

        Scope* s  = new Scope(scope);
        s->instance = i;

        return oi; // Stack handles iteration for subsequent instances
    }
    return nullptr;
}

} // namespace nucleoid
